Use the incomplete folder where appropriate

sledgehammer999 opened the pull request:
Restored torrents already have the correct path saved in their fastresume. Don't override it.
Magnets and new torrents have to be set explicitly to the incomplete folder if it is enabled.

Closes #11552

**Extra info**
The issue rose from a subtle different in `add_torrent_params` object between libtorrent 1.1.x and 1.2.x. In 1.1.x the object had an additiona field which held the fastresume data as whole. Decoding of the fastresume was done after we added the torrent in the libtorrent session. So some fields in fastresume were used instead of their counterparts in the `add_torrent_params` object. (and this could be overriden by supplying extra flags).
In 1.2.x the fastresume data are decoded and an `add_torrent_params` object is generated. The `add_torrent_params` object doesn't hold fastresume data as whole to be decoded afterwards. So we take the generated `add_torrent_params` object and adjust the values where appropriate.
For restored torrents setting the `save_path` field didn't take any effect for 1.1.x because it loaded the value from the embedded fastresume data. In 1.2.x it took effect and was pointed to the wrong path (aka the completed folder path).
This didn't manifest earlier because of a "weird" check in our code based on the presence of the `file_priority` field in the fastresume data. It was wrongly assumed that the field is always present in the fastresume. Apparently in a smaller number of cases the field wasn't present and "all hell broke loose".

sledgehammer999 commented:
@glassez @Chocobo1 please give review priority to this PR. If possible I would like to make a release this weekend.

glassez commented on src/base/bittorrent/session.cpp:
```diff
@@ -2160,6 +2134,27 @@ bool Session::addTorrent_impl(CreateTorrentParams params, const MagnetUri &magne
+    p.flags |= lt::torrent_flags::stop_when_ready;
+#endif
+
+    p.max_connections = maxConnectionsPerTorrent();
```
Why don't you allow to restore these two from resume data?

sledgehammer999 replied:
IMO, it is safer to always enforce our limits. (edge case:: the conf file has been edited manually)

glassez replied:
The question is "isn't possible to change these limits for particular torrent from qBittorrent?".

sledgehammer999 replied:
No. We expose only the ratio limits per torrent to the user. Changes to the max connections/uploads are done in bulk for all torrents.

Chocobo1 commented on src/base/bittorrent/session.cpp:
```diff
@@ -2058,27 +2060,21 @@ bool Session::addTorrent_impl(CreateTorrentParams params, const MagnetUri &magne
                 torrentInfo.stripRootFolder();
 
+            // If empty then Automatic mode, otherwise Manual mode
+            QString savePath = params.savePath.isEmpty() ? categorySavePath(params.category) : params.savePath;
```
`const QString savePath`

sledgehammer999 replied:
Read my code comment on the next line. `savePath` is adjusted by findIncomplete().

Chocobo1 commented on src/base/bittorrent/session.cpp:
```diff
@@ -2160,6 +2134,27 @@ bool Session::addTorrent_impl(CreateTorrentParams params, const MagnetUri &magne
         }
     }
 
+    // Common
```
If possible, I would like this "common" block go before `if()` conditionals, i.e. above line 2106

sledgehammer999 replied:
I keep it here because I want to bunch the similar if tests together.

glassez commented:
I would still recommend splitting addTorrent_impl into two variants (one for libtorrent-1.1.x and the second one for libtorrent-1.2.x) to make it easier to handle its logic in both cases (since it is quite different).
You can also return libtorrent-1.1 variant to its original (checked and reliable) form (as it is in 4.1.x branch).

sledgehammer999 commented:
@glassez I introduced the split between 1.1.x and 1.2.x. I followed what the v4.1.x branch does but I also added the changes I found here that are needed.

Chocobo1 commented on src/base/bittorrent/session.cpp:
```diff
@@ -1965,6 +1965,7 @@ bool Session::addTorrent(const TorrentInfo &torrentInfo, const AddTorrentParams
     return addTorrent_impl(CreateTorrentParams(params), MagnetUri(), torrentInfo);
 }
 
+#if (LIBTORRENT_VERSION_NUM < 10200)
```
why not move this inside the function?

sledgehammer999 replied:
done
